spec/light-client/Lightclient.tla

```diff
@@ -1,111 +1,273 @@
 ---------------------------- MODULE Lightclient ----------------------------
-EXTENDS Integers, Sequences, Bags
+(*
+ * A state-machine specification of the light client, following the English spec:
+ * https://github.com/tendermint/tendermint/blob/master/docs/spec/consensus/light-client.md
+ *
+ * Whereas the English specification presents the light client as a piece of sequential code,
+ * which contains non-tail recursion, we specify a state machine that explicitly has
+ * a stack of requests in its state. This specification can be easily extended to support
+ * multiple requests at the same time, e.g., to reduce latency.
+ *) 
 
-CONSTANT
-  Corr,     (* a set of correct nodes, which can act as validators *)
-  Faulty,   (* a set of faulty nodes, which can act as validators *)  
+EXTENDS Integers, Sequences
+
+\* the parameters of Lite Client
+CONSTANTS
+  TRUSTED_HEIGHT,
+    (* an index of the block header that the light client trusts by social consensus *)
+  TO_VERIFY_HEIGHT
+    (* an index of the block header that the light client tries to verify *)
+
+VARIABLES
+  state,        (* the current state of the light client *)
+  inEvent,      (* an input event to the light client, e.g., a header from a full node *)
+  outEvent,     (* an output event from the light client, e.g., finished with a verdict *)
+  requestStack,     (* the stack of requests to be issued by the light client to a full node *)
+  storedHeaders     (* the set of headers obtained from a full node*) 
```

Comment:
I feel we need some description of at least the elements of requestStack and storedHeaders, here or elsewhere. e.g. the request in the `requestStack` has `isLeft`, `startHeight`, `endHeight`. Headers have the block and the validator set. Also the `storedHeaders` behave like a set, there cannot be two headers with same height.

Reply:
It seems that Anca is also asking for types :) We need to figure out convincing way to "simulate" types as 1) that is common way engineers think about specs and 2) it is hard to follow spec without knowing what variables refer to.

Reply:
At a minimum we should explain in the comments.

Reply:
Short term we need to figure out how to take maximum out of TLA+. Mid/Long term I think we should explore alternatives, for example something like Stainless. As main users of this spec are engineers, we probably need something closer to the languages/tools they get used to.

Reply:
The standard practice in TLA+ is to write an invariant `TypeOK`, which basically checks variable types in every state. I will add one like that.

Moving to Stainless or similar tools will not solve all our problems, it will only give better assurance that the code matches the specifications. But this will come with the price of investing more efforts in writing pre-, post-conditions, and the invariants, as well as making them amenable for verification.

Reply:
See `TypeOK` in b06216bb54f8d26a8b1d6cc0123546dcbcb18957